lib/dry/operation/class_context/steps_method_prepender.rb

```diff
@@ -1,13 +1,32 @@
 # frozen_string_literal: true
 
+require "dry/operation/errors"
+
 module Dry
   class Operation
     module ClassContext
       # @api private
       class StepsMethodPrepender < Module
-        def initialize(method:)
+        FAILURE_HOOK_METHOD_NAME = :on_failure
+
+        RESULT_HANDLER = lambda do |result, instance, method|
+          return if result.success? || !instance.methods.include?(FAILURE_HOOK_METHOD_NAME)
```

Comment:
```suggestion
          return if result.success? || !(instance.methods.include?(FAILURE_HOOK_METHOD_NAME) || instance.private_methods.include?(FAILURE_HOOK_METHOD_NAME))
```
```suggestion
          return if result.success? || !instance.methods.include?(FAILURE_HOOK_METHOD_NAME)
```

Would it be worth doing this to allow `on_failure` to be private? If I'm thinking about the external interface to operation objects, certainly `#on_failure` doesn't feel like something we want to make callable externally.

Reply:
I also wonder if we could cache the result of this `methods`/`private_methods` lookup eventually...

Actually, this makes me wonder, instead of _looking up a possible on_failure method_, could we instead _implement_ a default empty `on_failure` method ourselves? We could define it on the instance whenever `StepsMethodPrepender` is included.

This would allow this result handler to be much more confident. It would mean this first line becomes a simpler `return if result.success?`.

Reply:
I changed it to use `#respond_to?` to make sure it works for any kind of method (public, private, inherited). What do you think? I also added a test to make sure it works with `private` methods.

> Actually, this makes me wonder, instead of looking up a possible on_failure method, could we instead implement a default empty on_failure method ourselves? We could define it on the instance whenever StepsMethodPrepender is included.

I'd like to have the `RESULT_HANDLER` injectable as it is now, this way we make all the magic in the gem configurable :/ WDYT?

Reply:
Thanks for incorporating private methods :)

I wasn't suggesting we change the injection of `RESULT_HANDLER`, instead what I was suggesting was this:

```
module Dry
  class Operation
    private

    def on_failure
      # empty method, always there by default
    end
  end
end
```

And inside `RESULT_HANDLER`, we no longer need to check for the presence of `#on_failure`, because we can rely on it always being there courtesy of the above. So this:

```ruby
        RESULT_HANDLER = lambda do |instance, method, result|
          return if result.success? ||
                    !(instance.methods + instance.private_methods).include?(
                      FAILURE_HOOK_METHOD_NAME
                    )
```

Becomes this:

```ruby
        RESULT_HANDLER = lambda do |instance, method, result|
          return if result.success?
```

Does that make sense? Create the method once, inside the gem code, and we no longer need to do introspection to find it.

I made those changes locally and all tests still passed.

---

While I'm here, I do have one question regarding `RESULT_HANDLER` being injectable. Who do you expect to be injecting alternatives? End users? Given of how fundamental that handler is to the behaviour of the library, do we actually _want_ it to ever be replaced? (In addition, `StepsMethodPrepender` is multiple layers removed from the classes that users are interacting with directly, anyway).

Reply:
> I wasn't suggesting we change the injection of RESULT_HANDLER, instead what I was suggesting was this:

However, in the case that another result handler was in use, it would make no sense that we're defining the `#on_failure` in the class.

> While I'm here, I do have one question regarding RESULT_HANDLER being injectable. Who do you expect to be injecting alternatives? End users? Given of how fundamental that handler is to the behaviour of the library, do we actually want it to ever be replaced? (In addition, StepsMethodPrepender is multiple layers removed from the classes that users are interacting with directly, anyway).

You're right, but `StepsMethodPrepender` can be completely disabled with `.skip_prepending`. I think it's good to avoid having any magic as something you can't, at least, opt-out. It's not only respectful with the user, but it also guides us through coming up with a better architecture. I think the `#on_failure` hook is the same case here. Let me know what you think and let's take a final decision 🙂

Reply:
@waiting-for-dev Thanks for explaining that, you've convinced me, happy to leave this as-is! 🚀

Reply:
@timriley, thanks for keeping an open mind here, your feedback is always so valuable! ❤️